File: DMBAlmanacProjectFolder/dmb-almanac-svelte/scraper/src/scrapers/guest-shows.ts
import { chromium, Page } from "playwright";
import * as cheerio from "cheerio";
import PQueue from "p-queue";
import { writeFileSync, readFileSync, existsSync } from "fs";
import { join, dirname } from "path";
import { randomDelay } from "../utils/rate-limit.js";
import { cacheHtml, getCachedHtml, saveCheckpoint, loadCheckpoint } from "../utils/cache.js";
import { normalizeWhitespace } from "../utils/helpers.js";

const BASE_URL = "https://www.dmbalmanac.com";
const OUTPUT_DIR = join(dirname(import.meta.url).replace("file://", ""), "../../output");

// Types for guest show appearances
export interface GuestShowAppearance {
  showId: string;
  showDate: string;
  venueName: string;
  city: string;
  state?: string;
  country: string;
  songs: GuestSongAppearance[];
}

export interface GuestSongAppearance {
  songTitle: string;
  songId?: string;
  instruments: string[];
  position?: number;
  set?: string;
}

export interface ScrapedGuestShows {
  guestId: string;
  guestName: string;
  totalAppearances: number;
  firstAppearanceDate?: string;
  lastAppearanceDate?: string;
  appearances: GuestShowAppearance[];
}

// Get all guest IDs from existing guest data
function getGuestIds(): { guestId: string; guestName: string }[] {
  const guestDetailsPath = join(OUTPUT_DIR, "guest-details.json");

  if (existsSync(guestDetailsPath)) {
    const data = JSON.parse(readFileSync(guestDetailsPath, "utf-8"));
    return data.guests.map((g: any) => ({
      guestId: g.originalId,
      guestName: g.name,
    }));
  }

  // Fallback to guests.json
  const guestsPath = join(OUTPUT_DIR, "guests.json");
  if (existsSync(guestsPath)) {
    const data = JSON.parse(readFileSync(guestsPath, "utf-8"));
    return data.guests.map((g: any) => ({
      guestId: g.originalId,
      guestName: g.name,
    }));
  }

  console.error("No guest data found!");
  return [];
}

// Parse TourGuestShows.aspx page for a specific guest
async function parseGuestShowsPage(
  page: Page,
  guestId: string,
  guestName: string
): Promise<ScrapedGuestShows | null> {
  const url = `${BASE_URL}/TourGuestShows.aspx?gid=${guestId}`;

  try {
    let html = getCachedHtml(url);

    if (!html) {
      await page.goto(url, { waitUntil: "networkidle", timeout: 30000 });
      html = await page.content();
      cacheHtml(url, html);
    }

    const $ = cheerio.load(html);

    const appearances: GuestShowAppearance[] = [];
    let firstDate: string | undefined;
    let lastDate: string | undefined;

    // Look for show entries - typically in a table format
    // TourGuestShows.aspx lists shows with dates, venues, and songs

    // Try finding show rows in table format
    $("tr").each((_, row) => {
      const $row = $(row);

      // Look for show links with TourShowSet.aspx
      const showLink = $row.find("a[href*='TourShowSet.aspx'][href*='id=']").first();
      if (!showLink.length) return;

      const href = showLink.attr("href") || "";
      const showIdMatch = href.match(/id=(\d+)/);
      if (!showIdMatch) return;

      const showId = showIdMatch[1];

      // Extract date - look for MM.DD.YYYY or YYYY-MM-DD format
      const rowText = $row.text();
      let showDate = "";

      // Try MM.DD.YYYY format first
      const dateMatch1 = rowText.match(/(\d{2})\.(\d{2})\.(\d{4})/);
      if (dateMatch1) {
        const [, month, day, year] = dateMatch1;
        showDate = `${year}-${month}-${day}`;
      } else {
        // Try YYYY-MM-DD format
        const dateMatch2 = rowText.match(/(\d{4})-(\d{2})-(\d{2})/);
        if (dateMatch2) {
          showDate = dateMatch2[0];
        }
      }

      // Track first/last dates
      if (showDate) {
        if (!firstDate || showDate < firstDate) firstDate = showDate;
        if (!lastDate || showDate > lastDate) lastDate = showDate;
      }

      // Extract venue info
      const venueLink = $row.find("a[href*='VenueStats.aspx']").first();
      const venueName = normalizeWhitespace(venueLink.text() || "");

      // Extract location - typically after venue name
      let city = "";
      let state = "";
      let country = "USA";

      const locationText = $row.text();
      const locationMatch = locationText.match(/([A-Za-z\s]+),\s*([A-Z]{2})(?:,\s*([A-Za-z\s]+))?/);
      if (locationMatch) {
        city = locationMatch[1].trim();
        state = locationMatch[2].trim();
        country = locationMatch[3]?.trim() || "USA";
      }

      // Extract songs played with this guest
      const songs: GuestSongAppearance[] = [];

      // Look for song links in the row or associated song section
      $row.find("a[href*='SongStats.aspx'][href*='sid='], a[href*='summary.aspx'][href*='sid=']").each((_, songEl) => {
        const songHref = $(songEl).attr("href") || "";
        const songIdMatch = songHref.match(/sid=(\d+)/);
        const songTitle = normalizeWhitespace($(songEl).text());

        if (songTitle) {
          // Look for instruments mentioned near this song
          const instruments: string[] = [];
          const songContext = $(songEl).parent().text();

          // Common instrument patterns
          const instrumentPatterns = [
            /guitar/i, /bass/i, /drums/i, /percussion/i, /keyboard/i, /piano/i,
            /violin/i, /viola/i, /cello/i, /trumpet/i, /saxophone/i, /sax/i,
            /harmonica/i, /vocals/i, /backing vocals/i, /lead vocals/i,
            /mandolin/i, /banjo/i, /fiddle/i, /flute/i, /clarinet/i
          ];

          for (const pattern of instrumentPatterns) {
            if (pattern.test(songContext)) {
              instruments.push(pattern.source.replace(/\\i$/, "").replace(/\\/g, ""));
            }
          }

          songs.push({
            songTitle,
            songId: songIdMatch ? songIdMatch[1] : undefined,
            instruments,
          });
        }
      });

      if (showId && (showDate || venueName)) {
        appearances.push({
          showId,
          showDate,
          venueName,
          city,
          state: state || undefined,
          country,
          songs,
        });
      }
    });

    // Alternative parsing: Look for list format
    if (appearances.length === 0) {
      $("li, .show-entry, .appearance").each((_, el) => {
        const $el = $(el);
        const showLink = $el.find("a[href*='TourShowSet.aspx'][href*='id=']").first();

        if (showLink.length) {
          const href = showLink.attr("href") || "";
          const showIdMatch = href.match(/id=(\d+)/);
          if (!showIdMatch) return;

          const showId = showIdMatch[1];
          const text = $el.text();

          // Parse date
          let showDate = "";
          const dateMatch = text.match(/(\d{2})\.(\d{2})\.(\d{4})/);
          if (dateMatch) {
            const [, month, day, year] = dateMatch;
            showDate = `${year}-${month}-${day}`;
          }

          if (showDate) {
            if (!firstDate || showDate < firstDate) firstDate = showDate;
            if (!lastDate || showDate > lastDate) lastDate = showDate;
          }

          // Parse venue
          const venueLink = $el.find("a[href*='VenueStats.aspx']").first();
          const venueName = normalizeWhitespace(venueLink.text() || showLink.text());

          appearances.push({
            showId,
            showDate,
            venueName,
            city: "",
            country: "USA",
            songs: [],
          });
        }
      });
    }

    // Get total count from page text if available
    let totalAppearances = appearances.length;
    const countMatch = $("body").text().match(/(\d+)\s+(?:total\s+)?(?:appearances?|shows?|performances?)/i);
    if (countMatch) {
      totalAppearances = parseInt(countMatch[1], 10);
    }

    return {
      guestId,
      guestName,
      totalAppearances,
      firstAppearanceDate: firstDate,
      lastAppearanceDate: lastDate,
      appearances,
    };
  } catch (error) {
    console.error(`Error parsing guest shows for ${guestName} (${guestId}):`, error);
    return null;
  }
}

// Main scraper function
export async function scrapeAllGuestShows(): Promise<ScrapedGuestShows[]> {
  console.log("Starting guest shows scraper...");
  console.log("This scraper extracts detailed appearance data from TourGuestShows.aspx pages");

  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage();

  await page.setExtraHTTPHeaders({
    "User-Agent": "DMBAlmanacClone/1.0 (Educational Project; Respectful Scraping)",
  });

  try {
    // Load checkpoint if exists
    const checkpoint = loadCheckpoint<{
      completedIds: string[];
      guestShows: ScrapedGuestShows[];
    }>("guest-shows");

    const completedIds = new Set(checkpoint?.completedIds || []);
    const allGuestShows: ScrapedGuestShows[] = checkpoint?.guestShows || [];

    // Get guest IDs to process
    const guests = getGuestIds();
    console.log(`Found ${guests.length} guests to process`);

    const remainingGuests = guests.filter((g) => !completedIds.has(g.guestId));
    console.log(`${remainingGuests.length} guests remaining to scrape`);

    // Rate limiter queue
    const queue = new PQueue({
      concurrency: 2,
      intervalCap: 5,
      interval: 10000,
    });

    let processed = 0;
    const total = remainingGuests.length;

    for (const guest of remainingGuests) {
      await queue.add(async () => {
        const guestShows = await parseGuestShowsPage(page, guest.guestId, guest.guestName);

        if (guestShows) {
          allGuestShows.push(guestShows);
          completedIds.add(guest.guestId);
          console.log(`  [${++processed}/${total}] ${guest.guestName}: ${guestShows.appearances.length} appearances`);
        } else {
          processed++;
          console.log(`  [${processed}/${total}] ${guest.guestName}: SKIPPED (no data)`);
        }

        await randomDelay(1000, 3000);
      });

      // Save checkpoint periodically
      if (processed % 25 === 0 && processed > 0) {
        saveCheckpoint("guest-shows", {
          completedIds: Array.from(completedIds),
          guestShows: allGuestShows,
        });
        console.log(`  [Checkpoint saved at ${processed}/${total}]`);
      }
    }

    await queue.onIdle();
    return allGuestShows;
  } finally {
    await browser.close();
  }
}

// Save guest shows to JSON file
export function saveGuestShows(guestShows: ScrapedGuestShows[]): void {
  const totalAppearances = guestShows.reduce((sum, g) => sum + g.appearances.length, 0);

  const output = {
    scrapedAt: new Date().toISOString(),
    source: BASE_URL,
    totalGuests: guestShows.length,
    totalAppearances,
    guestShows,
  };

  const filepath = join(OUTPUT_DIR, "guest-shows.json");
  writeFileSync(filepath, JSON.stringify(output, null, 2), "utf-8");
  console.log(`Saved ${guestShows.length} guests with ${totalAppearances} total appearances to ${filepath}`);
}

// Run if called directly
if (import.meta.url === `file://${process.argv[1]}`) {
  scrapeAllGuestShows()
    .then((guestShows) => {
      saveGuestShows(guestShows);
      console.log("\nDone!");
      console.log(`Total guests: ${guestShows.length}`);
      console.log(`Total appearances: ${guestShows.reduce((sum, g) => sum + g.appearances.length, 0)}`);
    })
    .catch((error) => {
      console.error("Scraper failed:", error);
      process.exit(1);
    });
}

File: projects/dmb-almanac/rust/crates/dmb_pipeline/src/scrape/guest_shows.rs
use anyhow::Result;
use dmb_core::Guest;
use scraper::Html;
use serde_json::{Value, json};

use super::{
    BASE_URL, ScrapeClient, normalize_whitespace, parse_i64_or_warn, parse_location, parse_mdy_any,
    regex, selector_or_warn, validate_required_fields, warn_if_empty, warn_if_no_selector_match,
    warn_if_out_of_range, warn_missing_field,
};

pub(super) fn scrape_guest_shows(client: &ScrapeClient, guests: &[Guest]) -> Result<Vec<Value>> {
    let mut entries = Vec::new();
    for guest in guests {
        if guest.id <= 0 {
            continue;
        }
        let url = format!("{BASE_URL}/TourGuestShows.aspx?gid={}", guest.id);
        let html = client.fetch_html(&url)?;
        let entry = parse_guest_shows_page(&html, guest.id, &guest.name);
        entries.push(entry);
    }
    Ok(entries)
}

pub(super) fn parse_guest_shows_page(html: &str, guest_id: i32, guest_name: &str) -> Value {
    let document = Html::parse_document(html);
    let mut appearances: Vec<Value> = Vec::new();
    let mut first_date: Option<String> = None;
    let mut last_date: Option<String> = None;

    let Some(row_selector) = selector_or_warn("tr", "tr") else {
        return empty_guest_shows_response(guest_id, guest_name, None, None);
    };
    let Some(show_link_selector) = selector_or_warn(
        "a[href*='TourShowSet.aspx'][href*='id=']",
        "a[href*='TourShowSet.aspx'][href*='id=']",
    ) else {
        return empty_guest_shows_response(guest_id, guest_name, None, None);
    };
    let Some(venue_selector) =
        selector_or_warn("a[href*='VenueStats.aspx']", "a[href*='VenueStats.aspx']")
    else {
        return empty_guest_shows_response(guest_id, guest_name, None, None);
    };
    let Some(song_selector) = selector_or_warn(
        "a[href*='SongStats.aspx'][href*='sid='], a[href*='summary.aspx'][href*='sid=']",
        "a[href*='SongStats.aspx'][href*='sid='], a[href*='summary.aspx'][href*='sid=']",
    ) else {
        return empty_guest_shows_response(guest_id, guest_name, None, None);
    };
    warn_if_no_selector_match(
        &document,
        "guest_shows",
        "showSelector",
        &[
            "a[href*='TourShowSet.aspx'][href*='id=']",
            "a[href*='ShowSetlist.aspx'][href*='id=']",
        ],
    );
    warn_if_no_selector_match(
        &document,
        "guest_shows",
        "venueSelector",
        &["a[href*='VenueStats.aspx']"],
    );
    warn_if_no_selector_match(
        &document,
        "guest_shows",
        "songSelector",
        &[
            "a[href*='SongStats.aspx'][href*='sid=']",
            "a[href*='summary.aspx'][href*='sid=']",
        ],
    );
    warn_if_empty(&document, &row_selector, "guest_shows", "tr");
    warn_if_empty(&document, &show_link_selector, "guest_shows", "show_link");
    warn_if_empty(&document, &venue_selector, "guest_shows", "venue_link");
    warn_if_empty(&document, &song_selector, "guest_shows", "song_link");

    for row in document.select(&row_selector) {
        let Some(show_link) = row.select(&show_link_selector).next() else {
            continue;
        };
        let show_id = show_link
            .value()
            .attr("href")
            .and_then(|href| regex(r"id=(\\d+)").captures(href))
            .and_then(|cap| cap.get(1).map(|m| m.as_str().to_string()));
        if show_id.as_deref().unwrap_or("").is_empty() {
            warn_missing_field("guest_shows", "showId");
        }
        let row_text = normalize_whitespace(&row.text().collect::<String>());
        let show_date = regex(r"(\\d{2}\\.[\\d]{2}\\.\\d{4}|\\d{4}-\\d{2}-\\d{2})")
            .captures(&row_text)
            .and_then(|cap| cap.get(1))
            .and_then(|m| {
                let value = m.as_str();
                parse_mdy_any(value).or_else(|| {
                    if regex(r"^\\d{4}-\\d{2}-\\d{2}$").is_match(value) {
                        Some(value.to_string())
                    } else {
                        None
                    }
                })
            });
        if let Some(date) = &show_date {
            if first_date.as_ref().is_none_or(|d| date < d) {
                first_date = Some(date.clone());
            }
            if last_date.as_ref().is_none_or(|d| date > d) {
                last_date = Some(date.clone());
            }
        }
        let venue_name = row
            .select(&venue_selector)
            .next()
            .map(|link| normalize_whitespace(&link.text().collect::<String>()))
            .unwrap_or_default();
        let (city, state, country) =
            parse_location(&row_text).unwrap_or((String::new(), None, "USA".to_string()));

        let mut songs: Vec<Value> = Vec::new();
        for song_link in row.select(&song_selector) {
            let song_title = normalize_whitespace(&song_link.text().collect::<String>());
            if song_title.is_empty() {
                continue;
            }
            let song_id = song_link
                .value()
                .attr("href")
                .and_then(|href| regex(r"sid=(\\d+)").captures(href))
                .and_then(|cap| cap.get(1).map(|m| m.as_str().to_string()));
            if song_id.as_deref().unwrap_or("").is_empty() {
                warn_missing_field("guest_shows", "songId");
            }
            let context_text = normalize_whitespace(&row_text).to_lowercase();
            let instruments = detect_instruments(&context_text);
            songs.push(json!({
                "songTitle": song_title,
                "songId": song_id,
                "instruments": instruments
            }));
        }
        appearances.push(json!({
            "showId": show_id,
            "showDate": show_date,
            "venueName": if venue_name.is_empty() { Value::Null } else { Value::String(venue_name) },
            "city": if city.is_empty() { Value::Null } else { Value::String(city) },
            "state": state,
            "country": country,
            "songs": songs
        }));
    }

    if appearances.is_empty() {
        let Some(li_selector) = selector_or_warn(
            "li, .show-entry, .appearance",
            "li, .show-entry, .appearance",
        ) else {
            return empty_guest_shows_response(guest_id, guest_name, first_date, last_date);
        };
        warn_if_empty(&document, &li_selector, "guest_shows", "li");
        for item in document.select(&li_selector) {
            let Some(show_link) = item.select(&show_link_selector).next() else {
                continue;
            };
            let show_id = show_link
                .value()
                .attr("href")
                .and_then(|href| regex(r"id=(\\d+)").captures(href))
                .and_then(|cap| cap.get(1).map(|m| m.as_str().to_string()));
            if show_id.as_deref().unwrap_or("").is_empty() {
                warn_missing_field("guest_shows", "showId");
            }
            let text = normalize_whitespace(&item.text().collect::<String>());
            let show_date = parse_mdy_any(&text);
            let venue_name = item.select(&venue_selector).next().map_or_else(
                || normalize_whitespace(&show_link.text().collect::<String>()),
                |link| normalize_whitespace(&link.text().collect::<String>()),
            );
            appearances.push(json!({
                "showId": show_id,
                "showDate": show_date,
                "venueName": if venue_name.is_empty() { Value::Null } else { Value::String(venue_name) },
                "city": Value::Null,
                "state": Value::Null,
                "country": "USA",
                "songs": []
            }));
        }
    }

    let total_appearances = regex(r"(\\d+)\\s+(?:total\\s+)?(?:appearances?|shows?|performances?)")
        .captures(&document.root_element().text().collect::<Vec<_>>().join(" "))
        .and_then(|cap| cap.get(1).map(|m| m.as_str()))
        .map_or_else(
            || i64::try_from(appearances.len()).unwrap_or(i64::MAX),
            |value| parse_i64_or_warn(Some(value), "guest_shows", "totalAppearances"),
        );
    if total_appearances > 0 {
        warn_if_out_of_range(
            "guest_shows",
            "totalAppearances",
            total_appearances,
            1,
            2000,
        );
    }

    if total_appearances > 0 && appearances.is_empty() {
        warn_missing_field("guest_shows", "appearances");
    }

    let mut result = json!({
        "guestId": guest_id,
        "guestName": guest_name,
        "totalAppearances": total_appearances,
        "firstAppearanceDate": first_date,
        "lastAppearanceDate": last_date,
        "appearances": appearances
    });
    validate_required_fields(
        "guest_shows",
        &result,
        &["guestId", "guestName", "totalAppearances", "appearances"],
    );
    if total_appearances > 0 {
        let empty = result
            .get("appearances")
            .and_then(|v| v.as_array())
            .is_none_or(std::vec::Vec::is_empty);
        if empty && let Some(obj) = result.as_object_mut() {
            obj.insert("_partial".to_string(), Value::Bool(true));
        }
    }
    result
}

fn empty_guest_shows_response(
    guest_id: i32,
    guest_name: &str,
    first_date: Option<String>,
    last_date: Option<String>,
) -> Value {
    json!({
        "guestId": guest_id,
        "guestName": guest_name,
        "totalAppearances": 0,
        "firstAppearanceDate": first_date,
        "lastAppearanceDate": last_date,
        "appearances": []
    })
}

fn detect_instruments(context_text: &str) -> Vec<String> {
    let mut instruments = Vec::new();
    let patterns = [
        "guitar",
        "bass",
        "drums",
        "percussion",
        "keyboard",
        "piano",
        "violin",
        "viola",
        "cello",
        "trumpet",
        "saxophone",
        "sax",
        "harmonica",
        "vocals",
        "backing vocals",
        "lead vocals",
        "mandolin",
        "banjo",
        "fiddle",
        "flute",
        "clarinet",
    ];
    for pattern in patterns {
        if context_text.contains(pattern) {
            instruments.push(pattern.to_string());
        }
    }
    instruments.sort();
    instruments.dedup();
    instruments
}
